### PR TITLE
use smart controls

### PR DESCRIPTION
**Pull request by ilBury**

### CODE REVIEW

File: webapp/view/ListReport.view.xml [hunk 1]
```diff
@@ -29,7 +31,7 @@
 					
 					<f:content>
 						
-						<fb:FilterBar id="filterbar" persistencyKey="myPersKey" useToolbar="false" >
+						<!-- <fb:FilterBar id="filterbar" persistencyKey="myPersKey" useToolbar="false" >
```
Comment (ho-thach-nguyen-lx) [inline webapp/view/ListReport.view.xml]: Remove comments if they're not needed

File: webapp/view/ListReport.view.xml [hunk 2]
```diff
@@ -3,6 +3,8 @@
 	xmlns:mvc="sap.ui.core.mvc"
 	xmlns:f="sap.f"
 	xmlns:core="sap.ui.core"
+	xmlns:smartTable="sap.ui.comp.smarttable"
+	xmlns:smartFilterBar="sap.ui.comp.smartfilterbar"
```
Comment (ho-thach-nguyen-lx) [inline webapp/view/ListReport.view.xml:6]: Do not forget to remove redundant namespaces.

File: webapp/view/ListReport.view.xml [hunk 3]
```diff
@@ -121,14 +123,40 @@
...
+							entitySet="Products" 
+						/>
 						
 					</f:content>
 				</f:DynamicPageHeader>
 			</f:header>
 				<f:content>
-				<Table
+					<smartTable:SmartTable 
+						entitySet="Products" enableAutoColumnWidth="true" enableExport="false"
```
Comment (ho-thach-nguyen-lx) [inline webapp/view/ListReport.view.xml]: Let's enable exporting, personalisation (aka p13n)

File: webapp/view/ListReport.view.xml [hunk 4]
```diff
@@ -121,14 +123,40 @@
...
+						enableAutoBinding="true" header="Products" useTablePersonalisation="false"
+						id="idSmartTable" smartFilterId="idSmartFilterBar" tableType="ResponsiveTable"
+						showTablePersonalisation="false" 
+					>
+						<smartTable:customToolbar>
+							<OverflowToolbar>
+								<ToolbarSpacer />
+								<Button
+									type="Emphasized"
+									text="{@i18n>CreateText}"
```
Comment (ho-thach-nguyen-lx) [inline webapp/view/ListReport.view.xml]: Don't use @i18n
It's intended to be used in annotations only. 
Tomorrow I will explain to you what the diff between i18n and @i18n is

File: webapp/annotations/annotation.xml [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+                            <Collection>
+                                <Record Type="Common.ValueListParameterInOut">
+                                    <PropertyValue Property="LocalDataProperty" PropertyPath="Name" />
+                                    <PropertyValue Property="ValueListProperty" String="Name" />
+                                </Record>
+                            </Collection>
+                        </PropertyValue>
+                    </Record>
+                </Annotation>
+                <Annotation Term="Common.ValueListWithFixedValues" />
```
Comment (ho-thach-nguyen-lx) [inline webapp/annotations/annotation.xml]: Let's make supplier filter as value help dialog, not combobox.
The reason why it shouldn't be combobox is there could be plenty suppliers. Therefore combobox is not suitable for this case, but value help dialog with table inside it.

Comment (ho-thach-nguyen-lx) [reply webapp/annotations/annotation.xml]: Combobox is intended to be used when you're sure there couldn't be more than 100 records and data structure is simple.
Category filter is the best case to use combobox but not suppliers.

**Review by ho-thach-nguyen-lx — COMMENTED** [review]
Red: add i18n 

Blue: Hide this field. You always need to hide technical fields from the end user.
They give them no info and sense

Снимок экрана 2024-04-03 в 16 03 21

**Review by ilBury — COMMENTED** [review]
As I said before I have a weird problem with select in objectPage and because of this we can't create/edit a new category and supplier with product.

**Comment by ho-thach-nguyen-lx** [comment]
Created At, Updated At should be read only

**Comment by ho-thach-nguyen-lx** [comment]
Hide column
Снимок экрана 2024-04-10 в 12 24 00

**Comment by ho-thach-nguyen-lx** [comment]
Add text
Снимок экрана 2024-04-10 в 12 24 57

File: webapp/controller/ListReport.controller.js [hunk 6]
```diff
@@ -63,182 +57,31 @@ sap.ui.define([
...
-
-		},
-
+	
 		onNavToObjectPage: function(oEvent) {
-			const oSource = oEvent.getSource();
-			const oCtx = oSource.getBindingContext();
+			const sNumberRegex = /[0-9]{1,}/gm;
+			const sPath = oEvent.getSource().getBindingContext().getPath();
+			const sProductId = sPath.match(sNumberRegex)[0];
```
Comment (ho-thach-nguyen-lx) [inline webapp/controller/ListReport.controller.js]: Get product id from the data, not path

File: webapp/controller/ProductDetails.controller.js [hunk 7]
```diff
@@ -94,11 +95,11 @@ sap.ui.define([
 				events: {
 					change: () => {
 						
-						const sSupplierId = this.getView().getBindingContext().getObject("Supplier")?.ID
-						this.getModel("FormModel").setProperty("/Supplier/ID", sSupplierId);
-						const sCategoryId = this.getView().getBindingContext().getObject("Category")?.ID
+						/* const sSupplierId = this.getView().getBindingContext().getObject("Supplier")?.ID
```
Comment (ho-thach-nguyen-lx) [inline webapp/controller/ProductDetails.controller.js]: Remove comments

File: webapp/controller/ProductDetails.controller.js [hunk 8]
```diff
@@ -111,27 +112,24 @@ sap.ui.define([
...
 			return [
-				this.byId("idNameInput"),
-				this.byId("idPriceInput"),
-				this.byId("idRatingInput"),
-				this.byId("idCategoriesSelect"),
-				this.byId("idSupplierNameSelect")
+				this.byId("idSmartNameField"),
+				this.byId("idSmartPriceField"),
+				this.byId("idSmartRatingField"),
+				/* this.byId("idSmartCategoryField"),
```
Comment (ho-thach-nguyen-lx) [inline webapp/controller/ProductDetails.controller.js]: comments

File: webapp/controller/ProductDetails.controller.js [hunk 9]
```diff
@@ -143,7 +141,7 @@ sap.ui.define([
 						message: "Should be required",
```
Comment (ho-thach-nguyen-lx) [inline webapp/controller/ProductDetails.controller.js]: i18n

File: webapp/controller/ProductDetails.controller.js [hunk 10]
```diff
@@ -154,25 +152,40 @@ sap.ui.define([
...
+				const sCategoryName = this.byId("idSmartCategoryField").getBindingContext().getObject("Category/Name");
+				
+				const sFilter = `/Categories/?$filter=Name eq '${sCategoryName}'`
+				this.getModel().update(path, {
+					uri: `https://services.odata.org/(S(3g0csrzc4YH3krwejcbcoqzz))/V2/OData/OData.svc/Categories(1)`		
+				}, {
+					"headers": {"Content-ID": 71}
+				})
+			}
+			/* const newCategoryURI = this.byId("idCategoriesSelect").getSelectedItem().getBindingContext().getPath();
```
Comment (ho-thach-nguyen-lx) [inline webapp/controller/ProductDetails.controller.js]: How many times i should repeat that no comments should left if they're meaningless?

File: webapp/controller/ProductDetails.controller.js [hunk 11]
```diff
@@ -181,9 +194,10 @@ sap.ui.define([
 			const aFormFields = this.getFormFields();
 			aFormFields.forEach(oControl => this.validateField(oControl));
 			const aMessages = Messaging.getMessageModel().getData();
+
+
```
Comment (ho-thach-nguyen-lx) [inline webapp/controller/ProductDetails.controller.js]: Why do we need these line breaks? It makes code review much harder.
If you want to beautify the code, use prettier. Automate what can be automated

File: webapp/controller/ProductDetails.controller.js [hunk 12]
```diff
@@ -221,19 +236,28 @@ sap.ui.define([
 		},
 
 		onCategoryChange: function(oControl) {
+			this.getModel().read("/Categories", {
+				success: (oData) => {
```
Comment (ho-thach-nguyen-lx) [inline webapp/controller/ProductDetails.controller.js]: Where is error handler?

**Comment by ho-thach-nguyen-lx** [comment]
Because of odata v2, smart controls and other things some files became redundant such as suppliers.json, fragments folder and so on.
Please delete them if they are truly redundant

File: webapp/view/ListReport.view.xml [hunk 13]
```diff
@@ -28,182 +27,47 @@
...
-							<cells>
-								<ObjectIdentifier title="{Name}"/>
-								<Text text="{Description}" />
-								<ObjectIdentifier title="{Price}"/>
-								
-								<Text 
-									text="{Supplier/Name}" 
+					<smartTable:SmartTable 
+						entitySet="Products" enableAutoColumnWidth="true"
+						enableAutoBinding="true" header="Products"
```
Comment (ho-thach-nguyen-lx) [inline webapp/view/ListReport.view.xml]: add i18n

File: webapp/view/ListReport.view.xml (hunk 4 above)
Comment (ho-thach-nguyen-lx) [reply webapp/view/ListReport.view.xml]: use @i18n in annotations only

**Review by ho-thach-nguyen-lx — CHANGES_REQUESTED** [review]

File: webapp/controller/ProductDetails.controller.js [hunk 14]
```diff
@@ -149,54 +147,58 @@ sap.ui.define([
...
+			const sPath = `/Products(${this.getView().getBindingContext().getObject("ID")})/$links/Category`
+			const sNewCategoryURI = this.byId("idCategoriesSelect").getSelectedItem().getBindingContext().getPath();
+			const bIsCreateMode = this.getModel("EditModel").getProperty("/CreateMode");
+
+			this.getModel().update(sPath, {
+				uri: `https://services.odata.org/(S(3g0csrzc4YH1lrwejcqcoqpz))/V2/OData/OData.svc/${sNewCategoryURI}`		
+			}, {
+				"headers": {"Content-ID": this.createNewId()},
+				success: () => {
+					if(bIsCreateMode) {
```
Comment (ho-thach-nguyen-lx) [inline webapp/controller/ProductDetails.controller.js:160]: where is error handler?

File: webapp/controller/ProductDetails.controller.js [hunk 15]
```diff
@@ -149,54 +147,58 @@ sap.ui.define([
...
-			})
+			const sPath = `/Products(${this.getView().getBindingContext().getObject("ID")})/$links/Supplier`
+			const sNewSupplierURI = this.byId("idSupplierNameSelect").getSelectedItem().getBindingContext().getPath();
+			const bIsCreateMode = this.getModel("EditModel").getProperty("/CreateMode");
 			
+			this.getModel().update(sPath, {
+				uri: `https://services.odata.org/(S(3g0csrzc4YH1lrwejcqcoqpz))/V2/OData/OData.svc/${sNewSupplierURI}`		
+			}, {
+				"headers": {"Content-ID": this.createNewId()},
+				success: () => {
```
Comment (ho-thach-nguyen-lx) [inline webapp/controller/ProductDetails.controller.js:179]: what will happen if the request fails?

File: webapp/manifest.json [hunk 16]
```diff
@@ -100,10 +111,13 @@
 					"useBatch": true,
 					"defaultBindingMode": "TwoWay",
 					"defaultCountMode": "Inline",
-					"disableHeadRequestForToken": true,
-					"groupId": "$auto"
+					"disableHeadRequestForToken": true
```
Comment (ho-thach-nguyen-lx) [inline webapp/manifest.json:114]: What does it solve?

**Review by ho-thach-nguyen-lx — CHANGES_REQUESTED** [review]

**Pull request merged**